src/util/logger/mod.rs

```diff
@@ -447,37 +360,120 @@ fn test_log_format() {
+        );
+
+        let duration = Duration::new(0, 123);
+        slog_debug!(logger, "Slow query";
+		    "sql" =>"SELECT * FROM TABLE WHERE ID=\"abc\"",
+		    "duration"=> ?duration,
+		    "process keys"=> 1500,
+	    );
+
+        slog_warn!(logger, "Type";
+            "Counter"=> ::std::f64::NAN,
+            "Score" => ::std::f64::INFINITY,
+            "Other" => ::std::f64::NEG_INFINITY);
+
+        let expect = r#"[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:469] []
+[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:469] [Welcome]
+[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:470] ["Welcome TiKV"]
+[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:471] [欢迎]
+[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:472] ["欢迎 TiKV"]
+[2019/01/15 13:40:39.615 +08:00] [INFO] [mod.rs:455] ["failed to fetch URL"] [backoff=3s] [attempt=3] [url=http://example.com]
+[2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:460] ["failed to \"fetch\" [URL]: http://example.com"]
+[2019/01/15 13:40:39.619 +08:00] [DEBG] [mod.rs:463] ["Slow query"] ["process keys"=1500] [duration=123ns] [sql="SELECT * FROM TABLE WHERE ID=\"abc\""]
+[2019/01/15 13:40:39.619 +08:00] [WARN] [mod.rs:473] [Type] [Other=-inf] [Score=inf] [Counter=NaN]
+"#;
+
+        BUFFER.with(|buffer| {
+            let buffer = buffer.borrow_mut();
+            let output = from_utf8(&*buffer).unwrap();
+
+            let re = Regex::new(r"(?P<datetime>\[.*?\])\s(?P<level>\[.*?\])\s(?P<source_file>\[.*?\])\s(?P<msg>\[.*?\])\s?(?P<kvs>\[.*\])?").unwrap();
+
+            for (output_line, expect_line) in output.lines().zip(expect.lines()) {
+                let expect_segments = re.captures(expect_line).unwrap();
+                let output_segments = re.captures(output_line).unwrap();
+
+                validate_log_datetime(peel(&expect_segments["datetime"]));
+
+                assert!(validate_log_source_file(
+                    peel(&expect_segments["source_file"]),
+                    peel(&output_segments["source_file"])
+                ));
+                assert_eq!(expect_segments["level"], output_segments["level"]);
+                assert_eq!(expect_segments["msg"], output_segments["msg"]);
+                assert_eq!(
+                    expect_segments.name("kvs").map(|s| s.as_str()),
+                    output_segments.name("kvs").map(|s| s.as_str())
+                );
+            }
+        });
+    }
+
+    /// Removes the wrapping signs, peels `"[hello]"` to `"hello"`, or peels `"(hello)"` to `"hello"`,
+    fn peel(output: &str) -> &str {
+        assert!(output.len() >= 2);
+        &(output[1..output.len() - 1])
+    }
+
+    /// Validates source file info.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// assert_eq!(true, validate_log_source_file("<unknown>", "<unknown>"));
+    /// assert_eq!(true, validate_log_source_file("mod.rs:1", "mod.rs:1"));
+    /// assert_eq!(true, validate_log_source_file("mod.rs:1", "mod.rs:100"));
```

Comment:
Why this is ok?

```
assert_eq!(true, validate_log_source_file("mod.rs:1", "mod.rs:100"));
```

Reply:
Once code in this file updated, the `date_time` and `sorce_file` fields of output will change, for example, from
> [2019/01/15 13:40:39.619 +08:00] [INFO] [mod.rs:1] [Welcome]  

to 

> [2019/01/16 11:21:31.121 +08:00] [INFO] [mod.rs:100] [Welcome]



This function just validates the `line number` is a valid `usize`, not asserts they are same value.

Reply:
ok